src/core/agent/AgentService.ts

```diff
@@ -272,10 +283,93 @@ export default class AgentService {
         }
         return dump
     }
+
+    async getCapabilities(token: string) {
+        if (token == this.#adminCredential) {
+            return [ALL_CAPABILITY]
+        }
+        
+        if (token === '') {
+            return [AUTH_CAPABILITY]
+        }
+
+        const key = this.getSigningKey()
+        let keyEncoder = new KeyEncoder('secp256k1')
+        const pemPublicKey = keyEncoder.encodePublic(key.publicKey, 'raw', 'pem')
+
+        const pubKeyObj = crypto.createPublicKey(pemPublicKey)
+        const { payload } = await jose.jwtVerify(token, pubKeyObj)
+
+        return payload.capabilities
+    }
+    
+    requestAuth(appName: string, appDesc: string, appUrl: string, capabilities: string) {
+        let requestId = uuidv4()
+        let authExtended = {
+            requestId,
+            auth: {
+                appName,
+                appDesc,
+                appUrl,
+                capabilities: JSON.parse(capabilities),
+            } as AuthInfo,
+        } as AuthInfoExtended
+        
+        this.#pubsub.publish(
+            PubSubInstance.EXCEPTION_OCCURRED_TOPIC,
+            {
+                title: "Request to authenticate application",
+                message: `${appName} is waiting for authentication, go to ad4min for more information.`,
+                type: ExceptionType.RequestAuth,
+                addon: JSON.stringify(authExtended),
+            } as ExceptionInfo
```

Comment:
If we generate the random number already here and send it to ad4min through this exception, we only need one more function to implement the [sketched out flow](https://hackmd.io/Epa06kVQQ5mkfB1i_Nqcqw) - that is we can do it without `permitAuth` below..

Reply:
...which also means, we wouldn't need `AuthInfoExtended` since we don't need to send the requestId to ad4min. We would need to send the random number with the `AuthInfo` though.

Reply:
This is tricky though, since user can inspect the content of this exception to get the reqestId and number, without actually confirm the request.